### ISSUE TITLE
New Rule Proposal: `no-unnecessary-setstate`

### ISSUE THREAD

**Issue by karlhorky** [yannickcr/eslint-plugin-react#2997]
**Rule proposal:** `no-unnecessary-setstate` - forbid setting state to the value of the existing state variable

Example of **incorrect** code:

```jsx
function Input (props) {
  const [name, setName] = useState('');	

  return (
    <input
      value={name}
      onChange={(event) => {
        setName(event.currentTarget.value);
        setName(name); // ESLint problem reported
      }}
    />
  );
}
```

Examples of **correct** code:

```jsx
function Input (props) {
  const [name, setName] = useState('');	

  return (
    <input
      value={name}
      onChange={(event) => {
        setName(event.currentTarget.value);
      }}
    />
  );
}
```

```jsx
function Input (props) {
  const [name, setName] = useState('');	

  return (
    <input
      value={name}
      onChange={(event) => {
        const name = event.currentTarget.value; // Shadowing
        setName(name);
      }}
    />
  );
}
```

**Comment by ljharb** [comment]
If it’s a common problem only with the useState hook, it seems like eslint-plugin-react-hooks would be a better place for it.

**Comment by karlhorky** [comment]
I'll try, although from experience, this probably won't get accepted over there.
